### DISCUSSION TITLE
automatic download of native libraries vs. missing internet-connection

### DISCUSSION THREAD

**Discussion by brainbytes42** [deepjavalibrary/djl#921] [Q&A]
Hello,

I'm just starting to evaluate DJL and, so far, it seems to be a great project. Thanks to anyone building this!

Reading the documentation, I stumbled upon this paragraph about GPU-Support: [https://d2l.djl.ai/chapter_installation/index.html#gpu-support](https://d2l.djl.ai/chapter_installation/index.html#gpu-support), which links to this part: [https://github.com/deepjavalibrary/djl/tree/master/mxnet/mxnet-engine#windows](https://github.com/deepjavalibrary/djl/tree/master/mxnet/mxnet-engine#windows).

Although this auto-detection sounds pretty comfortable, for me this seems to be a problem: 
I have two Windows machines - one is my usual development-machine, but without powerful hardware, and one is more powerful (2 GPUs), but has no internet-connection, not even via proxy. I can only reach the latter via LAN, but I need to transfer all gradle-dependencies by hand (or as fat-jar or whatsoever).

My question is now for advice how to find and provide the proper dependencies for the powerful offline-machine, especially those, that are native / auto-detected and as stated in the second link, not provided without auto-detection. (I could even copy my whole gradle-cache, but the weak online-machine would detect the wrong dependencies, as it has no GPU...)

Thanks in advance!

**Comment by frankfliu** [comment]
You can use PyTorch for windows if possible. PyTorch have windows offline package.

The MXNet for Windows GPU package is exceeded 2G limitation, we have to break it down for each cuda architecture. You can manually download them to your local. Here is the list of files:

For MXNet 1.8.0
```
https://publish.djl.ai/mxnet-1.8.0/win/common/libgcc_s_seh-1.dll.gz
https://publish.djl.ai/mxnet-1.8.0/win/common/libgfortran-3.dll.gz
https://publish.djl.ai/mxnet-1.8.0/win/common/libopenblas.dll.gz
https://publish.djl.ai/mxnet-1.8.0/win/common/libgcc_s_seh-1.dll.gz
https://publish.djl.ai/mxnet-1.8.0/win/common/libquadmath-0.dll.gz
https://publish.djl.ai/mxnet-1.8.0/win/cu102mkl/cudnn64_7.dll.gz
https://publish.djl.ai/mxnet-1.8.0/win/cu102mkl/mxnet_30.dll.gz
https://publish.djl.ai/mxnet-1.8.0/win/cu102mkl/mxnet_35.dll.gz
https://publish.djl.ai/mxnet-1.8.0/win/cu102mkl/mxnet_37.dll.gz
https://publish.djl.ai/mxnet-1.8.0/win/cu102mkl/mxnet_50.dll.gz
https://publish.djl.ai/mxnet-1.8.0/win/cu102mkl/mxnet_52.dll.gz
https://publish.djl.ai/mxnet-1.8.0/win/cu102mkl/mxnet_60.dll.gz
https://publish.djl.ai/mxnet-1.8.0/win/cu102mkl/mxnet_61.dll.gz
https://publish.djl.ai/mxnet-1.8.0/win/cu102mkl/mxnet_70.dll.gz
https://publish.djl.ai/mxnet-1.8.0/win/cu102mkl/mxnet_75.dll.gz
https://publish.djl.ai/mxnet-1.8.0/win/cu110mkl/mxnet_50.dll.gz
https://publish.djl.ai/mxnet-1.8.0/win/cu110mkl/mxnet_60.dll.gz
https://publish.djl.ai/mxnet-1.8.0/win/cu110mkl/mxnet_70.dll.gz
https://publish.djl.ai/mxnet-1.8.0/win/cu110mkl/mxnet_75.dll.gz
https://publish.djl.ai/mxnet-1.8.0/win/cu110mkl/mxnet_80.dll.gz
```

For MXNet 1.7.0:
```
https://publish.djl.ai/mxnet-1.7.0-backport/win/common/libgcc_s_seh-1.dll.gz
https://publish.djl.ai/mxnet-1.7.0-backport/win/common/libgfortran-3.dll.gz
https://publish.djl.ai/mxnet-1.7.0-backport/win/common/libopenblas.dll.gz
https://publish.djl.ai/mxnet-1.7.0-backport/win/common/libgcc_s_seh-1.dll.gz
https://publish.djl.ai/mxnet-1.7.0-backport/win/common/libquadmath-0.dll.gz
https://publish.djl.ai/mxnet-1.7.0-backport/win/cu102mkl/cudnn64_7.dll.gz
https://publish.djl.ai/mxnet-1.7.0-backport/win/cu102mkl/mxnet_30.dll.gz
https://publish.djl.ai/mxnet-1.7.0-backport/win/cu102mkl/mxnet_35.dll.gz
https://publish.djl.ai/mxnet-1.7.0-backport/win/cu102mkl/mxnet_37.dll.gz
https://publish.djl.ai/mxnet-1.7.0-backport/win/cu102mkl/mxnet_50.dll.gz
https://publish.djl.ai/mxnet-1.7.0-backport/win/cu102mkl/mxnet_52.dll.gz
https://publish.djl.ai/mxnet-1.7.0-backport/win/cu102mkl/mxnet_60.dll.gz
https://publish.djl.ai/mxnet-1.7.0-backport/win/cu102mkl/mxnet_61.dll.gz
https://publish.djl.ai/mxnet-1.7.0-backport/win/cu102mkl/mxnet_70.dll.gz
https://publish.djl.ai/mxnet-1.7.0-backport/win/cu102mkl/mxnet_75.dll.gz
```

DJL check local cache folder first. If you manually downloaded the required dll copy to your offline machine:
1. Put the required .dll into default cache location
2. or use MXNET_LIBRARY_PATH environment variable to set the location

Note: 
1. All the files in common directory are required. cudnn64_7.dll.gz is required for cu102.
2. You need gunzip the files into .dll
3. You only need download the matching cuda architecture, and rename it to mxnet.dll

**Reply by brainbytes42** [reply]
Thank you very much - so now I know there is a solution, then I can first learn more about the library and then try out your solution. :-)

**Comment by frankfliu** [comment]
There actually a better solution if you know your target cuda architecture
1. modify build.gradle downloadMxnetNativeLib task https://github.com/deepjavalibrary/djl/blob/master/mxnet/native/build.gradle#L169 to download windows GPU librarys
2. run the following command:
```
gradle downloadMxnetNativeLib
gradle publish
```

You will find a jar for windows gpu in build/repo/... folder. Then you can include this jar in your project, and everything should just work.
